### ISSUE TITLE
Ecto.Changeset.no_assoc_constraint/3 does not work

### ISSUE THREAD

**Issue by snofang** [elixir-sqlite/ecto_sqlite3#124]
**Test environment:**
```
Erlang/OTP 25 [erts-13.1.2] 
Elixir 1.14.2 
ecto 3.10.3
ecto_sql 3.10.1
ecto_sqlite3 0.10.3
```
**Description:**
Suppose it, as  a completely straightforward use case, as the documentation of `no_assoc_constraint/3` expressed (and the same code is working fine on postgrex).

P.S. As I am reporting this, I see there are two other releases come out and sounds like non of them addressed this issue, so I  didn't check them.

**Comment by warmwaffles** [comment]
> sounds like non of them addressed this issue, so I didn't check them.

That is correct, none of them address this. We'll need to add unit tests for this and figure out why it was missed. This was a feature added 8 years ago to ecto.

**Comment by warmwaffles** [comment]
Looks like a test would have covered this but we ignore it because we can't get nicely formatted FK errors https://github.com/elixir-sqlite/ecto_sqlite3/blob/37b68d9874695e5c17b80c5d7d57f8a803fd9524/integration_test/test_helper.exs#L85-L88

**Comment by thangpham7793** [comment]
It's possible to get details on the FKs of a table using `PRAGMA foreign_key_list(table-name)`. Would it be possible/desirable to call that here to return the contraint name?

https://github.com/elixir-sqlite/ecto_sqlite3/blob/794efccfc4f7f98df46fea6ffc937087ff17ff42/lib/ecto/adapters/sqlite3/connection.ex#L150-L153

**Comment by jakeprem** [comment]
It seems this also applies to `Ecto.Changeset.foreign_key_constraint/3`

**Comment by warmwaffles** [comment]
Yea it's mainly having to figure out how to capture that error from SQLite.

**Comment by jakeprem** [comment]
As a quick workaround I added the following to my repo module:
```elixir
def safe_insert(changeset, opts \\ []) do
    insert(changeset, opts)
  rescue
    e in Ecto.ConstraintError ->
      handle_foreign_key_constraints(changeset, e)
  end

defp handle_foreign_key_constraints(%Ecto.Changeset{} = changeset, e) do
    %{constraints: constraints} = changeset
    foreign_key_constraints = Enum.filter(constraints, &(&1.type == :foreign_key))

    if e.type == :foreign_key and foreign_key_constraints != [] do
      error_changeset =
        Enum.reduce(foreign_key_constraints, changeset, fn constraint, changeset ->
          Ecto.Changeset.add_error(
            changeset,
            constraint.field,
            "may not exist (SQLite raised a foreign key constraint error but does not specify which one)"
          )
        end)

      {:error, error_changeset}
    else
      raise e
    end
end
```
There's likely a better workaround but hopefully that's helpful for anyone else running into this.

**Comment by warmwaffles** [comment]
I'm not going to close this issue. When I get some time I'll fiddle around with the error outputs and see if we can't get more details about the failure to pick out the failing constraint. There has to be a way to make SQLite explain which one blew up.
